### PR TITLE
Add comprehensive CI workflows

### PR DESCRIPTION
**Pull request by Ragarnoy**

### CODE REVIEW

File: .github/workflows/check.yml [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+# the new branch. Restricting the push trigger to the main branch ensures that the PR only gets
+# built once.
+on:
+  push:
+    branches: [main]
+  pull_request:
+# If new code is pushed to a PR branch, then cancel in progress workflows for that PR. Ensures that
+# we don't waste CI time, and returns results quicker https://github.com/jonhoo/rust-ci-conf/pull/5
+concurrency:
+  group: ${{ github.workflow }}-${{ github.head_ref || github.run_id }}
```
Comment (elpiel) [inline .github/workflows/check.yml:20]: I usually use git_ref (iirc) not sure about head_ref and run_id. Hopefully they 2 stop commits to the same branch building at the same time and building only the last one

**Review by elpiel — CHANGES_REQUESTED** [review]

**Comment by elpiel** [comment]
Is there a reason for so many workflows and jobs? I believe optimizing for the build (with codecov, lints, tests, etc.) and a release workflows would be ideal

**Comment by Ragarnoy** [comment]
> Is there a reason for so many workflows and jobs? I believe optimizing for the build (with codecov, lints, tests, etc.) and a release workflows would be ideal

compartmentalizing mostly, I like to keep tasks separate

File: .github/workflows/check.yml [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+    name: stable / fmt
+    steps:
+      - uses: actions/checkout@v4
+        with:
+          submodules: true
+      - name: Install stable
+        uses: dtolnay/rust-toolchain@stable
+        with:
+          components: rustfmt
+      - name: cargo fmt --check
```
[the remark this answers is not in the document]
Comment (Ragarnoy) [reply .github/workflows/check.yml:35]: yeah i know it's just so that it's easy to spot in the workflow

File: .github/codecov.yml [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+        paths: ["src/blocking.rs"]
+      nonblocking:
+        target: 85%
+        paths: ["src/nonblocking.rs"]
+
+# Make comments less noisy but informative
+comment:
+  layout: "files"
+  require_changes: true
+  show_carryforward_flags: true
```
Comment (Ragarnoy) [inline .github/codecov.yml:36]:
```suggestion
  show_carryforward_flags: true
  
```

File: .github/workflows/check.yml (hunk 1 above)
Comment (Ragarnoy) [reply .github/workflows/check.yml:20]: if it does not behave as expected i'll use your solution

File: .github/workflows/check.yml [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+        run: cargo fmt --check
+  clippy:
```
[the remark this answers is not in the document]
Comment (Ragarnoy) [reply .github/workflows/check.yml:37]: I'll run them on beta for now ik it's better to use nightly on those tools but i'd rather never get those false positives

**Pull request merged**